src/services/codefixes/importFixes.ts

```diff
@@ -118,561 +144,546 @@ namespace ts.codefix {
-        function createCodeAction(
-            description: DiagnosticMessage,
-            diagnosticArgs: string[],
-            changes: FileTextChanges[],
-            kind: ImportCodeActionKind,
-            moduleSpecifier?: string): ImportCodeAction {
-            return {
-                description: formatMessage.apply(undefined, [undefined, description].concat(<any[]>diagnosticArgs)),
-                changes,
-                kind,
-                moduleSpecifier
-            };
+        return getCodeActionForImport(symbol, { ...context, symbolName, kind: ImportKind.Namespace });
+    }
+
+    function getActionsForNonUMDImport(context: ImportCodeFixContext, allSourceFiles: ReadonlyArray<SourceFile>, cancellationToken: CancellationToken): ImportCodeAction[] {
+        const { sourceFile, checker, symbolName, symbolToken } = context;
+        // "default" is a keyword and not a legal identifier for the import, so we don't expect it here
+        Debug.assert(symbolName !== "default");
+        const symbolIdActionMap = new ImportCodeActionMap();
+        const currentTokenMeaning = getMeaningFromLocation(symbolToken);
+
+        eachOtherExternalModule(checker, allSourceFiles, sourceFile, moduleSymbol => {
+            cancellationToken.throwIfCancellationRequested();
+            // check the default export
+            const defaultExport = checker.tryGetMemberInModuleExports("default", moduleSymbol);
+            if (defaultExport) {
+                const localSymbol = getLocalSymbolForExportDefault(defaultExport);
+                if (localSymbol && localSymbol.escapedName === symbolName && checkSymbolHasMeaning(localSymbol, currentTokenMeaning)) {
+                    // check if this symbol is already used
+                    const symbolId = getUniqueSymbolId(localSymbol, checker);
+                    symbolIdActionMap.addActions(symbolId, getCodeActionForImport(moduleSymbol, { ...context, kind: ImportKind.Default }));
+                }
+            }
+
+            // check exports with the same name
+            const exportSymbolWithIdenticalName = checker.tryGetMemberInModuleExportsAndProperties(symbolName, moduleSymbol);
+            if (exportSymbolWithIdenticalName && checkSymbolHasMeaning(exportSymbolWithIdenticalName, currentTokenMeaning)) {
+                const symbolId = getUniqueSymbolId(exportSymbolWithIdenticalName, checker);
+                symbolIdActionMap.addActions(symbolId, getCodeActionForImport(moduleSymbol, { ...context, kind: ImportKind.Named }));
+            }
+        });
+
+        return symbolIdActionMap.getAllActions();
+    }
+
+    function checkSymbolHasMeaning({ declarations }: Symbol, meaning: SemanticMeaning): boolean {
+        return some(declarations, decl => !!(getMeaningFromDeclaration(decl) & meaning));
+    }
+
+    export function eachOtherExternalModule(checker: TypeChecker, allSourceFiles: ReadonlyArray<SourceFile>, sourceFile: SourceFile, cb: (module: Symbol) => void) {

```

Comment:
the name is rather confusing, not sure what it does really from reading it, why not `forEachExternalModule`? and the caller can exclude the current file.

Reply:
it is called in two places, so i would just check the current sourcefile is not the one in the list at the call sites